### PR TITLE
Platform Versioning

### PR DESCRIPTION
**Pull request by guilhermebr**
related to https://github.com/tsuru/tsuru/issues/2036

### CODE REVIEW

File: app/image/platform.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	"github.com/tsuru/tsuru/log"
+)
+
+type platformImages struct {
+	PlatformName string `bson:"_id"`
+	Images       []string
+	Count        int
+}
+
+func platformImagesColl() (*storage.Collection, error) {
```
[the remark this answers is not in the document]
Comment (guilhermebr) [reply app/image/platform.go]: For sure, I will move this and `app/images.go` DB things to `storage/mongodb` in a separated commit

**Comment by cezarsa** [comment]
We must not forget to eventually update the image GC to be aware of versioned platform images: https://github.com/tsuru/tsuru/blob/0b4c233f1bcd430551870631a69fba7dfbc0f3ba/app/image/gc/gc.go

File: app/image/platform.go [hunk 2]
```diff
@@ -7,107 +7,74 @@ package image
...
-		ReturnNew: true,
-		Upsert:    true,
-	}
-	_, err = coll.FindId(platformName).Apply(dbChange, &imgs)
-	if err != nil {
-		return "", err
-	}
-	return fmt.Sprintf("%s/%s:v%d", basicImageName("tsuru"), platformName, imgs.Count), nil
+type platformImages struct {
+	Name   string `bson:"_id"`
```
Comment (ggarnier) [inline app/image/platform.go]: You don't need this `bson` tag here

File: storage/mongodb/platform_image.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	dbStorage "github.com/tsuru/tsuru/db/storage"
+	"github.com/tsuru/tsuru/types/app/image"
+)
+
+var _ image.PlatformImageStorage = &PlatformImageStorage{}
+
+type PlatformImageStorage struct{}
+
+type platformImage struct {
+	Name   string `bson:"_id"`
```
Comment (ggarnier) [inline storage/mongodb/platform_image.go]: I personally don't like using values like names as collection ids. Specially in cases like this, where it's the name of another entity (platform). I'd prefer using the platform name as a regular field, and creating an index for it.

**Comment by guilhermebr** [comment]
@ggarnier PTAL

**Comment by guilhermebr** [comment]
@cezarsa PTAL

**Pull request merged**